### PR TITLE
#254 - temp file removal and stream closure in RqMultipart

### PR DESCRIPTION
**Pull request by exper0**
#254 - temp file removal and stream closure in RqMultipart

### CODE REVIEW

**Comment by exper0** [comment]
@davvd please find reviewer for this one

**Comment by davvd** [comment]
@exper0 Thanks, let me find someone who can review this pull request

**Comment by davvd** [comment]
@pinaf it's yours, please review

**Comment by exper0** [comment]
@pinaf please hold on with review, it doesn't work as expected, I'll make one more fix

**Comment by exper0** [comment]
@pinaf now it's ready for review

File: src/main/java/org/takes/rq/TempInputStream.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+package org.takes.rq;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Input stream using a temporary cache file.
+ *
+ * <p>All implementations of this interface must be immutable and thread-safe.
```
Comment (pinaf) [inline src/main/java/org/takes/rq/TempInputStream.java]: @exper0 this is not an interface

File: src/main/java/org/takes/rq/RqMultipart.java [hunk 2]
```diff
@@ -230,18 +230,23 @@ public Base(final Request req) throws IOException {
...
             final File file = File.createTempFile(
                 RqMultipart.class.getName(), ".tmp"
             );
-            file.deleteOnExit();
             final FileChannel channel = new RandomAccessFile(
                 file, "rw"
             ).getChannel();
             try {
                 channel.write(
-                    ByteBuffer.wrap(this.head().iterator().next().getBytes())
```
Comment (pinaf) [inline src/main/java/org/takes/rq/RqMultipart.java]: @exper0 this is unnecessary

File: src/main/java/org/takes/rq/TempInputStream.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    /**
+     * Original stream.
+     */
+    private final transient InputStream origin;
+
+    /**
+     * Temporary file used as a cache.
+     */
+    private final transient File file;
```
Comment (pinaf) [inline src/main/java/org/takes/rq/TempInputStream.java]: @exper0 what is this file used for? Isn't the point of this class to write to the file as it reads from the stream? right now the only thing you do with the file is close it.

File: src/main/java/org/takes/rq/CapInputStream.java [hunk 4]
```diff
@@ -34,6 +34,10 @@
  * @author Yegor Bugayenko (yegor@teamed.io)
  * @version $Id$
  * @since 0.16
+ * @todo #254:30min CapInputStream should delegate all standard InputStream
```
Comment (pinaf) [inline src/main/java/org/takes/rq/CapInputStream.java]: @exper0 why do you need this puzzle if you are no longer using `CapInputStream` in `RqMultipart`?

File: src/main/java/org/takes/rq/RqMultipart.java (hunk 2 above)
Comment (exper0) [inline src/main/java/org/takes/rq/RqMultipart.java]: @pinaf do you mean additional newlines?

File: src/main/java/org/takes/rq/TempInputStream.java (hunk 3 above)
Comment (exper0) [inline src/main/java/org/takes/rq/TempInputStream.java]: @pinaf exactly - the file is needed is just to close it. See the original idea in discussion #254 - it was proposed by @yegor256. The idea is just to have ability to remove temporary file on stream `close`

File: src/main/java/org/takes/rq/RqMultipart.java (hunk 2 above)
Comment (pinaf) [inline src/main/java/org/takes/rq/RqMultipart.java]: @exper0 yeah there was no logic change, right? no need to break it.

File: src/main/java/org/takes/rq/TempInputStream.java (hunk 3 above)
Comment (pinaf) [inline src/main/java/org/takes/rq/TempInputStream.java]: @exper0 Ok in that case let's change the javadoc for this class to make it clear that all it does is delete the given file on close - otherwise it looks like the stream is actually backed by the file in the sense of the data coming in.

File: src/main/java/org/takes/rq/CapInputStream.java (hunk 4 above)
Comment (exper0) [inline src/main/java/org/takes/rq/CapInputStream.java]: @pinaf hmm, I thought I answered to this... my comment is missing. The idea is to remove temp files on body stream close. There should quite common logic - after reading request body there should `close` on the stream However, in some cases body InputStream can be wrapped with CapInputStream - so in this case we just get IOException which is wrong. That's why I created this puzzle.

**Comment by exper0** [comment]
@pinaf changed java doc and repaired indentation.

File: src/main/java/org/takes/rq/TempInputStream.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+ * SOFTWARE.
+ */
+package org.takes.rq;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Input stream wrapper that removed associated File instance on close.
```
Comment (pinaf) [inline src/main/java/org/takes/rq/TempInputStream.java]: @exper0 `removed` -> `removes`

File: src/test/java/org/takes/rq/TempInputStreamTest.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * TempInputStream can delete the underlying temporary file.
+     * @throws IOException if some problem occurs.
+     */
+    @Test
+    public void deletesTempFile() throws IOException {
+        final File file = File.createTempFile("tempfile", ".tmp");
+        final BufferedWriter out = new BufferedWriter(new FileWriter(file));
+        out.write("Temp file deletion test");
+        out.close();
```
Comment (pinaf) [inline src/test/java/org/takes/rq/TempInputStreamTest.java]: @exper0 let's use try/finally blocks for closing instances of Closeable

File: src/test/java/org/takes/rq/TempInputStreamTest.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        final InputStream body = new TempInputStream(
+            new FileInputStream(file), file
+        );
+        body.close();
```
Comment (pinaf) [inline src/test/java/org/takes/rq/TempInputStreamTest.java]: @exper0 same here

File: src/test/java/org/takes/rq/TempInputStreamTest.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        MatcherAssert.assertThat(
```
Comment (pinaf) [inline src/test/java/org/takes/rq/TempInputStreamTest.java]: @exper0 let's also add an assert that the file exists before we close the stream

**Comment by pinaf** [comment]
@exper0 thanks, 4 more and we are good.

File: src/test/java/org/takes/rq/TempInputStreamTest.java (hunk 7 above)
Comment (exper0) [inline src/test/java/org/takes/rq/TempInputStreamTest.java]: @pinaf what should I put into try/finally block here? In above cases there at least write, but here we have only `body` creation and immediate `close` on it.

Comment (pinaf) [inline src/test/java/org/takes/rq/TempInputStreamTest.java]: @exper0 we just need to make sure that close is called always - that's why we need a `finally` block

Comment (pinaf) [inline src/test/java/org/takes/rq/TempInputStreamTest.java]: @exper0 can we move both calls to close into a single finally block?

**Comment by exper0** [comment]
@pinaf done.

File: src/test/java/org/takes/rq/TempInputStreamTest.java (hunk 7 above)
Comment (exper0) [inline src/test/java/org/takes/rq/TempInputStreamTest.java]: @pinaf inserted additional assert between creation and close, see new commit

**Comment by pinaf** [comment]
@exper0 do you think we can have a single `finally` block with the 2 calls to `close`?

**Comment by exper0** [comment]
@pinaf if we need absolutely reliable call to `close` in both cases when we need two blocks. We can put something in finally block only if the object we're closing was created before try block. Or we need to use null references and check them in finally block. Do you see what I mean?

**Comment by pinaf** [comment]
@rultor merge

File: src/main/java/org/takes/rq/TempInputStream.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * Closes the Input stream, deleting associated file.
+     * @throws IOException if some problem occurs.
+     */
+    @Override
+    public void close() throws IOException {
+        super.close();
+        this.origin.close();
+        if (!this.file.delete()) {
+            throw new IOException("unable to delete file");
```
Comment (yegor256) [inline src/main/java/org/takes/rq/TempInputStream.java]: @exper0 see http://www.yegor256.com/2015/12/01/rethrow-exceptions.html

**Comment by yegor256** [comment]
@exper0 see above

**Comment by exper0** [comment]
@yegor256 changed it, see new commit

**Comment by pinaf** [comment]
@yegor256 I have a feeling @yegor256 was referring to adding context to the exception being thrown.

**Comment by exper0** [comment]
@pinaf there is no context - it's generic class that remove some file on stream close. The class doesn't know any wider context

**Comment by exper0** [comment]
@yegor256 please take a look into updated code

File: src/main/java/org/takes/rq/RqMultipart.java [hunk 10]
```diff
@@ -251,9 +254,9 @@ private Request make(final byte[] boundary) throws IOException {
             }
             return new RqWithHeader(
                 new RqLive(
-                    new CapInputStream(
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqMultipart.java]: @exper0 why did you remove `CapInputStream`?

**Comment by yegor256** [comment]
@exper0 see above. I didn't understand why `CapInputStream` is removed.

File: src/main/java/org/takes/rq/RqMultipart.java (hunk 10 above)
Comment (exper0) [inline src/main/java/org/takes/rq/RqMultipart.java]: @yegor256 CapInputStream is not needed for files, the stream knows file length without CapInputStream

**Comment by yegor256** [comment]
@rultor merge

**Comment by davvd** [comment]
@ypshenychka review this ticket please, for compliance with our [quality rules](http://at.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor deploy pls

**Comment by ypshenychka** [comment]
@davvd Quality is good.

**Comment by davvd** [comment]
> @davvd Quality is good.

@ypshenychka thank you, good quality is always good :)

**Comment by davvd** [comment]
@pinaf just added 10 mins to @ypshenychka (for QA), payment ID is `76573022`; **31 mins** sent to your balance (ID `AP-7B7515052B006335Y`), many thanks! It took 304 hours and 54 mins.; extra minutes for review comments (c=16); added +31 to your rating, now it is equal to [+6800](http://www.netbout.com/b/35128?open=rating)
